Firestore/core/src/firebase/firestore/remote/serializer.cc

```diff
@@ -1492,101 +1493,202 @@ google_firestore_v1_MapValue Serializer::EncodeMapValue(
   return result;
 }
 
-// TODO(varconst): everything below
-
 MutationResult Serializer::DecodeMutationResult(
+    nanopb::Reader* reader,
     const google_firestore_v1_WriteResult& write_result,
     const SnapshotVersion& commit_version) const {
-  // TODO(varconst): implement.
-  // - (MutationResult)decodedMutationResult:(GCFSWriteResult *)mutation
-  //     commitVersion:(const SnapshotVersion &)commitVersion;
-  return MutationResult{{}, {}};
+  // NOTE: Deletes don't have an update_time, use commit_version instead.
+  SnapshotVersion version =
+      write_result.has_update_time
+          ? DecodeSnapshotVersion(reader, write_result.update_time)
+          : commit_version;
+  absl::optional<std::vector<FieldValue>> transform_results;
+  if (write_result.transform_results_count > 0) {
+    transform_results = std::vector<FieldValue>{};
+    for (pb_size_t i = 0; i < write_result.transform_results_count; i++) {
+      transform_results->push_back(
+          DecodeFieldValue(reader, write_result.transform_results[i]));
+    }
+  }
+  return MutationResult(std::move(version), std::move(transform_results));
 }
 
 std::unordered_map<std::string, std::string>
 Serializer::EncodeListenRequestLabels(const QueryData& query_data) const {
-  // TODO(varconst): implement.
-  // - (nullable NSMutableDictionary<NSString *, NSString *> *)
-  //     encodedListenRequestLabelsForQueryData: (const QueryData &)queryData;
-  return {};
+  auto value = EncodeLabel(query_data.purpose());
+
+  std::unordered_map<std::string, std::string> result(1);
+  if (!value.empty()) {
+    result["goog-listen-tags"] = value;
+  }
+  return result;
 }
 
 std::string Serializer::EncodeLabel(QueryPurpose purpose) const {
-  // TODO(varconst): implement.
-  // - (nullable NSString *)encodedLabelForPurpose:(QueryPurpose)purpose;
-  return {};
+  switch (purpose) {
+    case QueryPurpose::Listen:
+      return "";
+    case QueryPurpose::ExistenceFilterMismatch:
+      return "existence-filter-mismatch";
+    case QueryPurpose::LimboResolution:
+      return "limbo-document";
+    default:
```

Comment:
Nit: when a switch is exhaustive, rather than adding a `default` case, it's better to add the `UNREACHABLE` macro _after_ the switch (see examples in this file). The advantage is that the compiler will issue a warning if there are unhandled cases in the switch, but that warning is canceled out by a `default` case.